kong/tracing/instrumentation.lua

```diff
@@ -0,0 +1,198 @@
+local pdk_tracer = require "kong.pdk.tracing".new()
+local time_ns    = require "kong.tools.utils".time_ns
+local tablepool  = require "tablepool"
+local tablex     = require "pl.tablex"
+local hooks      = require "kong.hooks"
+
+local table         = table
+local pack          = table.pack
+local unpack        = table.unpack
+
+local instrument_tracer = pdk_tracer
+local NOOP = function() end
+
+local instrumentations = {}
+local available_types = {}
+
+-- db query
+function instrumentations.db_query(connector)
+  local f = connector.query
+
+  local function wrap(self, sql, ...)
+    local span = instrument_tracer.start_span("query", {
+      attributes = {
+        sql = sql,
+      }
+    })
+    local ret = pack(f(self, sql, ...))
+    span:finish()
+    return unpack(ret)
+  end
+
+  connector.query = wrap
+end
+
+-- router
+function instrumentations.router()
+  hooks.register_hook("runloop:access:router:pre", function (ctx)
+    return instrument_tracer.start_span("router")
+  end)
+
+  hooks.register_hook("runloop:access:router:post", function (span, match_t)
+    if not span then
+      return
+    end
+
+    span:finish()
+  end)
+end
+
+-- http_request (root span)
+-- we won't set the propagation headers there to avoid conflict with other tracing plugins
+function instrumentations.http_request()
+  local req = kong.request
+
+  local method = req.get_method()
+  local path = req.get_path()
+  local span_name = method .. " " .. path
+
+  local start_time = ngx.ctx.KONG_PROCESSING_START
```

Comment:
Question: since you've spent some time working on this area of the codebase and thinking about these kinds of problems, what do you think about this module (or something else within the `kong.tracing` namespace) eventually becoming the source of truth for all of the `ngx.ctx.KONG_*` latency data instead of merely consuming it? I think it would be nice to replace all of [this](https://github.com/Kong/kong/blob/c5fd7233ab3e0d910fef31bc3eecd8bdc25fe220/kong/init.lua#L1298-L1377) with a couple API calls to another module.

[Here](https://github.com/Kong/kong/pull/8688#discussion_r850872162) is one such discussion I had recently that got me thinking about this.

Reply:
I think it's reasonable and more suitable to make a `kong.time` / `kong.resty.time` module to handle all timing related stuff, like converting time between precisions and fetching cached phase execution times for convenience and better code structure.

I also wonder if we can get more accurate phase execution times using OpenResty/Nginx's API.